File: day1/typescript/day1.ts
import { readFile } from 'fs';
import { promisify } from 'util';

import * as R from 'ramda';

const rotate = <T>(dist: number, list: T[]) => (
    R.flatten<T>(R.reverse(R.splitAt(dist, list)))
);

const sumPairs = (dist: number, list: number[]) => (
    R.compose(
        R.sum,
        R.zipWith<number, number, number>(
            (a, b) => a === b ? a : 0,
            rotate(dist, list),
        ),
    )(list)
);

async function day1() {
    const input = await promisify(readFile)('day1/input.txt', 'utf8');
    const digits = R.map(parseInt, Array.from(input));

    const part1 = sumPairs(1, digits);
    const part2 = sumPairs(digits.length / 2, digits);

    console.log(part1);
    console.log(part2);
}

day1();

File: day1/haskell/day1.hs
rotate :: Int -> [a] -> [a]
rotate _ [] = []
rotate n xs = zipWith const (drop n (cycle xs)) xs

sumPairs :: Int -> [Int] -> Int
sumPairs d xs = sum . zipWith fn (rotate d xs) $ xs
    where fn a b = if a == b then a else 0

solve :: Int -> String -> Int
solve d = sumPairs d . map (read . pure)

main :: IO ()
main = do
    input <- readFile "day1/input.txt"
    print . solve 1 $ input
    print . solve (length input `div` 2) $ input